Add Bitbucket authentication topic (#1788)
codenvy/che-docs: modules/administration-guide/partials/proc_configuring_bitbucket_authentication.adoc

// configuring-bitbucket-authentication

[id="configuring_bitbucket_authentication_{context}"]
= Authentication on Bitbucket servers

{prod} users may use public or private Bitbucket Git repositories as a source of theirs projects. The standard

> factory flow using devfile at the root of the repository is available starting of {version} of {prod}.

Reviewer: {version} is not displayed in https://5ffeb9e4243f44caf45604e6--eclipse-che-docs-pr.netlify.app/che-7/administration-guide/authenticating-users-3rd-party-services/
please check that it's the right property name

Using private repositories, requires some additional configuration described below.

Bitbucket authentication is based on using so-called "personal access tokens". Each Bitbucket user is able to request some
amount of personal access tokens with different names, permissions, expiration times and so on. Those tokens
may be used to sign Bitbucket REST API calls and perform Git repository operations.

> To allow Bitbucket authentication on {prod} side, personal tokens must be stored in user's namespace in a form of

Reviewer: It's in the administration guide but actually there is nothing admin can do with it for users.

Maybe `Configuring Bitbucket servers` should be a part of Administration but `Authentication on Bitbucket servers` end-user guide?

Author: Not sure if users can  create secret with cretential for themselves ?

Reviewer: Do you mean that RBAC don't allow them to create secrets?
I checked and in case of OSIO, users are not able to CRUD secrets in che related namespaces(suffixed with -che), but they are able in their personal. But OSIO is a special case. I don't see any security reason in allowing users creating their secrets, if they are not - how then they can configure their credentials for docker registry?...

But at the same time is admin really is supposed to create those secrets? I don't think SCM admin has access to other users personal tokens.
Is no - who is supposed to create these secrets?

Author: I believe this feature will never be enabled on OSIO, it's purely for custom CRW installations. And i don't know if the CRW users is able to access to something more than theirs UD (i mean K8S/Openshift console/CLI/etc) to manage their own namespaces and create secrets in them. It's looks like more admin stuff as for me.

Reviewer 2: I agree with @sleshchenko, how an admin is supposed to get the token of another user? I don't think that's possible. And as for other secrets in the user namespace, those can/should be managed by the user. Ideally users should only see UD, that's why we are automating the creation of those secrets, but we are not there yet.

Author: So how the users supposed to create secrets for themself now ?

Reviewer 2: We have a full documentation section for that: https://www.eclipse.org/che/docs/che-7/end-user-guide/mounting-a-secret-as-a-file-or-an-environment-variable-into-a-workspace-container/. That's where I was expecting to find the documentation for the bitbucket token too.

I agree that all those secrets should be managed from the UD, not manually with `kubectl`. We have [an issue for that](https://github.com/eclipse/che/issues/17954). But we are not there yet and anyway those should not be included among admin tasks (admins do not manage users git token, ssh keypairs etc...)

secret. The secret must look as follows:

[source,yaml]
----
include::example$snip_bitbucket-personal-access-token-secret.adoc[]
----

The main parts of the secret are:

[cols=2*]
|===
| `app.kubernetes.io/component`

> | Label indicates it is a SCM personal token secret

Reviewer: Is the value supposed to be always `scm-personal-access-token` ?

Reviewer: Please consider specifying at least once that SCM is Source Code Management

Author: Added mention

| `che.eclipse.org/che-userid`
| {prod} id of the user token belongs to

| `che.eclipse.org/scm-userid`

> | SCM id of the user token belongs to

Reviewer: Is user supposed to create scm userid and username? What if these values are changed and the specified token belongs to another user?

Author: rephrased

Reviewer: I still don't understand if it's a user/admin who supposed to set userID, expiration time... I'm not familiar with access token on bitbucket but github UI provides me only personal access token (I assume bitbucket does the same, does not provide userid or expiration time for token)

> | `che.eclipse.org/scm-username`

Reviewer: Are all of these parts labels except `token` or only `app.kubernetes.io/component` while other are expected as secret items?

Reviewer: Ah, from the example above I got that they are annotations, but I think this table should be self-sufficient

Author: I added the type column

// configuring-bitbucket-authentication

[id="configuring_bitbucket_authentication_{context}"]
= Authentication on Bitbucket servers

> {prod} users may use public or private repositories Bitbucket SCM (Source Code Management) system  as a source of theirs projects. The standard

Reviewer 3: ```suggestion
{prod} users may use public or private repositories Bitbucket SCM (Source Code Management) system as a source of their projects. The standard
```

factory flow using devfile at the root of the repository is available starting of 7.25 version of {prod}.

> Using private repositories, requires some additional configuration described below.

Reviewer 3: ```suggestion
The use of private repositories requires some additional configuration described below.
```

> Bitbucket authentication is based on using so-called "personal access tokens". Each Bitbucket user is able to request some

Reviewer 3: ```suggestion
Bitbucket authentication is based on using personal access tokens. Each Bitbucket user is able to request some
```

> amount of personal access tokens with different names, permissions, expiration times and so on. Those tokens

Reviewer 3: ```suggestion
amount of personal access tokens with different names, permissions, expiration times, and so on. Those tokens
```

> may be used to sign Bitbucket REST API calls and perform Git repository operations.

Reviewer 3: ```suggestion
can be used to sign Bitbucket REST API calls and perform Git repository operations.
```

To allow Bitbucket authentication on {prod} side, personal tokens must be stored in user's namespace in a form of
secret. The secret must look as follows:

[source,yaml]
----
include::example$snip_bitbucket-personal-access-token-secret.adoc[]
----

The main parts of the secret are:

[cols=3*]
|===
| Label
| `app.kubernetes.io/component`
| Indicates it is a SCM personal token secret. 

| Annotation
| `che.eclipse.org/che-userid`
| {prod} id of the user token belongs to

| Annotation
| `che.eclipse.org/scm-userid`
| Bitbucket user id to which token belongs

| Annotation
| `che.eclipse.org/scm-username`
| Bitbucket user name to which token belongs

| Annotation
| `che.eclipse.org/scm-url`
| Bitbucket server URL to which this token belong

| Annotation
| `che.eclipse.org/expired-after`
| Personal access token expiration time

| Data entry
| `token`
| Base-64 encoded value of the personal access token

|===

> Using token credential provided in this secret, another one is automatically created to allow authorized Git operations. It will be mount

Reviewer 3: ```suggestion
Using the token credential provided in this secret, another one is automatically created to allow authorized Git operations. It will be mount
```
